File: MockLibAndroid/SmartMockLib/src/test/java/com/crescentflare/smartmock/utility/SmartMockParamMatcherTest.java
package com.crescentflare.smartmock.utility;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Assert;
import org.junit.Test;

/**
 * Utility test: parameter matching cases
 */
public class SmartMockParamMatcherTest
{
    /**
     * Test cases
     */

    @Test
    public void testParamEquals() throws Exception
    {
        Assert.assertTrue(SmartMockParamMatcher.paramEquals("username", "username"));
        Assert.assertFalse(SmartMockParamMatcher.paramEquals("username", "username10"));
        Assert.assertTrue(SmartMockParamMatcher.paramEquals("*name", "username"));
        Assert.assertFalse(SmartMockParamMatcher.paramEquals("*not", "username"));
        Assert.assertTrue(SmartMockParamMatcher.paramEquals("*@*", "user@mail"));
        Assert.assertTrue(SmartMockParamMatcher.paramEquals("user??*@*.*", "user10ex@mail.com"));
        Assert.assertFalse(SmartMockParamMatcher.paramEquals("user??*@*.*", "user4@mail.com"));
        Assert.assertTrue(SmartMockParamMatcher.paramEquals("user??*@*.*", "user42@mail.com"));
    }

    @Test
    public void testDeepEquals() throws Exception
    {
        Assert.assertTrue(SmartMockParamMatcher.deepEquals(
                makeObject("{ 'username': 'username' }"),
                makeObject("{ 'username': 'username', 'extra': 'value' }")
        ));
        Assert.assertTrue(SmartMockParamMatcher.deepEquals(
                makeObject("{ 'username': 'username', 'info': { 'name': '*', 'role': 'admin' } }"),
                makeObject("{ 'username': 'username', 'info': { 'name': 'test', 'role': 'admin' } }")
        ));
        Assert.assertFalse(SmartMockParamMatcher.deepEquals(
                makeObject("{ 'username': 'username', 'info': { 'name': '*', 'role': 'admin' } }"),
                makeObject("{ 'username': 'username', 'info': { 'role': 'admin' } }")
        ));
    }


    /**
     * Helper
     */

    private JSONObject makeObject(String jsonString)
    {
        jsonString = jsonString.replaceAll("'", "\"");
        try
        {
            return new JSONObject(jsonString);
        }
        catch (JSONException ignored)
        {
        }
        return new JSONObject();
    }
}

File: MockLibIOS/Example/Tests/SmartMockParamMatcherTest.swift
//
//  SmartMockParamMatcherTest.swift
//  SmartMockLib Tests
//
//  Utility test: parameter matching cases
//

import UIKit
import XCTest
@testable import SmartMockLib

class SmartMockParamMatcherTest: XCTestCase {
    
    func testParamEquals() {
        XCTAssertTrue(SmartMockParamMatcher.paramEquals(requireParam: "username", haveParam: "username"))
        XCTAssertFalse(SmartMockParamMatcher.paramEquals(requireParam: "username", haveParam: "username10"))
        XCTAssertTrue(SmartMockParamMatcher.paramEquals(requireParam: "*name", haveParam: "username"))
        XCTAssertFalse(SmartMockParamMatcher.paramEquals(requireParam: "*not", haveParam: "username"))
        XCTAssertTrue(SmartMockParamMatcher.paramEquals(requireParam: "*@*", haveParam: "user@mail"))
        XCTAssertTrue(SmartMockParamMatcher.paramEquals(requireParam: "user??*@*.*", haveParam: "user10ex@mail.com"))
        XCTAssertFalse(SmartMockParamMatcher.paramEquals(requireParam: "user??*@*.*", haveParam: "user4@mail.com"))
        XCTAssertTrue(SmartMockParamMatcher.paramEquals(requireParam: "user??*@*.*", haveParam: "user42@mail.com"))
    }
    
    func testDeepEquals() {
        XCTAssertTrue(SmartMockParamMatcher.deepEquals(
            requireDictionary: [ "username": "username" as AnyObject ],
            haveDictionary: [ "username": "username" as AnyObject, "extra": "value" as AnyObject ]
        ))
        XCTAssertTrue(SmartMockParamMatcher.deepEquals(
            requireDictionary: [ "username": "username" as AnyObject, "info": [ "name": "*", "role": "admin" ] as AnyObject ],
            haveDictionary: [ "username": "username" as AnyObject, "info": [ "name": "test", "role": "admin" ] as AnyObject ]
        ))
        XCTAssertFalse(SmartMockParamMatcher.deepEquals(
            requireDictionary: [ "username": "username" as AnyObject, "info": [ "name": "*", "role": "admin" ] as AnyObject ],
            haveDictionary: [ "username": "username" as AnyObject, "info": [ "role": "admin" ] as AnyObject ]
        ))
    }
    
}
